### DISCUSSION TITLE
Many 'generated' languages in settings

### DISCUSSION THREAD

**Discussion by keunes** [WeblateOrg/weblate#7039] [General]
Hello,

I just ended up on my profile, and saw for example the following on the 'languages' tab:
![afbeelding](https://user-images.githubusercontent.com/11229646/147891372-aafaf94b-f4cc-4b88-a529-f9d70f367a13.png)

And there's plenty of other seemingly random entries that include 'generated'.

I have no clue what they are, but they certainly don't seem to be languages. Shouldn't they be deleted from the database, or otherwise at least hidden from those lists of languages?

**Comment by nijel** [comment]
These are languages that matched the file mask in the component configuration, see https://docs.weblate.org/en/latest/admin/languages.html#parsing-language-codes for info how that works.

PS: If these were created by an error, you can remove unused ones by `weblate cleanup_languages --delete`.

**Reply by keunes** [reply]
Thanks for explaining & the reference. I'm reporting this from what I see on hosted Weblate.

<details><summary>When searching for 'generated', I can see the following list of 'languages'</summary>

```
ajp (generated) (ajp)
appendix (generated) (appendix)
appetite (generated) (appetite)
apps (generated) (apps)
biodentis (generated) (biodentis)
c1 (generated) (c1)
c2 (generated) (c2)
cadstar (generated) (cadstar)
cares (generated) (cares)
classes (generated) (classes)
controlflow (generated) (controlflow)
datastructures (generated) (datastructures)
dry (generated) (dry)
dwosle (generated) (dwosle)
envfiles (generated) (envfiles)
errors (generated) (errors)
fl (generated) (fl)
floatingpoint (generated) (floatingpoint)
* (generated) (*)
😃 (generated) (😃)
index (generated) (index)
inputoutput (generated) (inputoutput)
interactive (generated) (interactive)
interpreter (generated) (interpreter)
introduction (generated) (introduction)
laserdenta (generated) (laserdenta)
locale (generated) (locale)
lyra (generated) (lyra)
lzh (generated) (lzh)
microdent (generated) (microdent)
modules (generated) (modules)
night (generated) (night)
opera (generated) (opera)
qz_MM (generated) (qz_MM)
round (generated) (round)
scanlab (generated) (scanlab)
shera (generated) (shera)
stdlib2 (generated) (stdlib2)
stdlib (generated) (stdlib)
straumann (generated) (straumann)
template (generated) (template)
untranslated (generated) (untranslated)
va (generated) (va)
venv (generated) (venv)
whatnow (generated) (whatnow)
will_PAGINATE (generated) (will_PAGINATE)
zeno (generated) (zeno)
```
</p>
</details>

Some people clearly made some mistake, and 'polluted' the language list. I can imagine that 'va' or 'ajp' are actual languages. But things like 'template', 'untranslated', '🙂', 'whatnow' and 'c1' are very unlikely to be languages.

For AntennaPod I have disabled that everyone can add languages so there's no problem for me. But it does clutter the interface for translators and might be taking some time from projects if a user accidentally added such nonsensical entry as a language. I think it would be good, therefore, to clean these up. I wouldn't mind reaching out to project maintainers who have any of these languages in their project, to check if that was intentional and ask them to delete it from their project (I assume a language must not be used before it gets deleted by the cleanup script).

**Reply by nijel** [reply]
Indeed, some people messed up the structure: https://hosted.weblate.org/projects/python-docs-ar/python-docs-ar/

Those languages should not be available for users to select, the list is filtered since 099f2d3d857.

**Reply by nijel** [reply]
I've just added definition of some missing languages to https://github.com/WeblateOrg/language-data/ to cover some generated ones. This is something we do time to time to catch weirdly named locales used in the wild.
